parsely/src/main/java/com/parsely/parselyandroid/EngagementManager.java

```diff
@@ -0,0 +1,114 @@
+package com.parsely.parselyandroid;
+
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TimeZone;
+import java.util.Timer;
+import java.util.TimerTask;
+
+/**
+ * Engagement manager for article and video engagement.
+ * <p>
+ * Implemented to handle its own queuing of future executions to accomplish
+ * two things:
+ * <p>
+ * 1. Flushing any engaged time before canceling.
+ * 2. Progressive backoff for long engagements to save data.
+ */
+class EngagementManager {
+
+    private final ParselyTracker parselyTracker;
+    public Map<String, Object> baseEvent;
+    private boolean started;
+    private final Timer parentTimer;
+    private TimerTask waitingTimerTask;
+    private long latestDelayMillis, totalTime;
+    private Calendar startTime;
+    private final UpdateEngagementIntervalCalculator intervalCalculator;
+
+    public EngagementManager(ParselyTracker parselyTracker, Timer parentTimer, long intervalMillis, Map<String, Object> baseEvent, UpdateEngagementIntervalCalculator intervalCalculator) {
```

Comment:
Minor (🔍): This line exceed the usual `120` character limit guideline that we have on other projects, consider reformatting it to something more readable like:

```java
    public EngagementManager(
            ParselyTracker parselyTracker,
            Timer parentTimer,
            long intervalMillis,
            Map<String, Object> baseEvent,
            UpdateEngagementIntervalCalculator intervalCalculator
    ) {
```

Reply:
Sure, but I'd leave this, and other static code changes, to some linter. I'd prefer to not focus on this on reviews, as I'm afraid it'll shadow other, more functional comments. Do you have a suggestion for a linter/formatter tool for Java?

Reply:
> Sure, but I'd leave this, and other static code changes, to some linter. I'd prefer to not focus on this on reviews, as I'm afraid it'll shadow other, more functional comments.

Sure, this is just a minor one, feel free to skip those when needed. 👍 

> Do you have a suggestion for a linter/formatter tool for Java?

On most of our other Java-utilizing repos, we are and have been using `CheckStyle`, but then again we are not actively updating this tool on a regular basis, or at all for that matter, thus I am not sure. Having said that, `CheckStyle` indeed seems to be the best we can do for now, in only for consistency purpose... 🤷

Reply:
I've fixed the comments! After the second thought, it's probably not worth to introduce Java linter - we plan to use Kotlin anyway.

Reply:
> I've fixed the comments! 

Checking @wzieba and thanks! 👀 

> After the second thought, it's probably not worth to introduce Java linter - we plan to use Kotlin anyway.

Yeaaa, maybe not... 🤷